pkg/ipcache/ipcache.go

```diff
@@ -360,6 +366,7 @@ func (ipc *IPCache) upsertLocked(
 				newIdentity.shadowed = true
 			}
 		}
+		ipc.prefixLengths.Add([]netip.Prefix{cidrCluster.AsPrefix()})
```

Comment:
nit: I'd prefer it if we extracted the prefixes into a local variable, and only add it `ipc.prefixLengths` once we actually start inserting the entry into the data structures. 

That way, if we ever add an early return before here and the actual upsert (currently on line 407), we don't do any early mutations. There is no bug at the moment, but given that the code up until here did not mutate any of `ipc` I think it would be nice to keep that.

Reply:
Applying this diff on top:

```
$ git dc
diff --git a/pkg/ipcache/ipcache.go b/pkg/ipcache/ipcache.go
index cd2c966f999c..c9162c7e2e23 100644
--- a/pkg/ipcache/ipcache.go
+++ b/pkg/ipcache/ipcache.go
@@ -369,7 +369,6 @@ func (ipc *IPCache) upsertLocked(
                                newIdentity.shadowed = true
                        }
                }
-               ipc.prefixLengths.Add([]netip.Prefix{cidrCluster.AsPrefix()})
        } else if addrCluster, err := cmtypes.ParseAddrCluster(ip); err == nil { // Endpoint IP or Endpoint IP with ClusterID
                cidrCluster = addrCluster.AsPrefixCluster()

@@ -392,7 +391,6 @@ func (ipc *IPCache) upsertLocked(
                                }
                        }
                }
-               ipc.prefixLengths.Add([]netip.Prefix{cidrCluster.AsPrefix()})
        } else {
                log.WithFields(logrus.Fields{
                        logfields.AddrCluster: ip,
@@ -420,6 +418,7 @@ func (ipc *IPCache) upsertLocked(
                ipc.identityToIPCache[newIdentity.ID] = map[string]struct{}{}
        }
        ipc.identityToIPCache[newIdentity.ID][ip] = struct{}{}
+       ipc.prefixLengths.Add([]netip.Prefix{cidrCluster.AsPrefix()})

        if hostIP == nil {
                delete(ipc.ipToHostIPCache, ip)
@@ -624,7 +623,6 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
                        scopedLog.Debug("Deleting CIDR shadowed by endpoint IP")
                        callbackListeners = false
                }
-               ipc.prefixLengths.Delete([]netip.Prefix{cidrCluster.AsPrefix()})
        } else if addrCluster, err := cmtypes.ParseAddrCluster(ip); err == nil { // Endpoint IP or Endpoint IP with ClusterID
                // Convert the endpoint IP into an equivalent full CIDR.
                cidrCluster = addrCluster.AsPrefixCluster()
@@ -647,7 +645,6 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
                                callbackListeners = false
                        }
                }
-               ipc.prefixLengths.Delete([]netip.Prefix{cidrCluster.AsPrefix()})
        } else {
                scopedLog.Error("Attempt to delete invalid IP from ipcache layer")
                metrics.IPCacheErrorsTotal.WithLabelValues(
@@ -665,6 +662,7 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
        }
        delete(ipc.ipToHostIPCache, ip)
        delete(ipc.ipToK8sMetadata, ip)
+       ipc.prefixLengths.Delete([]netip.Prefix{cidrCluster.AsPrefix()})

        // Update named ports
        namedPortsChanged = false
```